### PR TITLE
feat: DBTP-1503 - Terraform codebase pipelines

### PR DESCRIPTION
**Pull request by JohnStainsby**
Terraform pipelines based on the `platform-config.yml`. Each pipeline has a codebuild job to create deploy manifests and deploy actions for each service. The deploy order is set by run groups in the config.

Example section from platform-config.yml:
```yml
codebase_pipelines:
  - name: test
    repository: uktrade/demodjango
    services:
      - run_group_1:
        - web
      - run_group_2:
        - api
        - celery-worker
        - celery-beat
    pipelines:
      - name: main
        branch: main
        environments:
          - name: dev
```

Created pipeline:

image

Addresses most of https://uktrade.atlassian.net/browse/DBTP-1503.
Manual release pipeline and prod trigger pipeline will be separate PRs.

### CODE REVIEW

File: codebase-pipelines/buildspec-manifests.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+version: 0.2
+
+env:
+  ${jsonencode({
+    exported-variables: flatten([ for env in environments : ["CLUSTER_NAME_${env}", [ for svc in services: "SERVICE_NAME_${env}_${svc}" ]]])
+  })}
```
Comment (WillGibson) [inline codebase-pipelines/buildspec-manifests.yml:6]: Does this work? Might be a bit easier to follow. Not a deal breaker though.
```suggestion
  ${jsonencode({
    exported-variables: flatten([
      for env in environments : [
        "CLUSTER_NAME_${env}",
        [for svc in services: "SERVICE_NAME_${env}_${svc}"]
      ]
    ])
  })}
```

File: codebase-pipelines/codebuild.tf [hunk 2]
```diff
@@ -113,3 +113,78 @@ resource "aws_codebuild_webhook" "codebuild_webhook" {
...
+  logs_config {
+    cloudwatch_logs {
+      group_name  = aws_cloudwatch_log_group.codebase_deploy_manifests.name
+      stream_name = aws_cloudwatch_log_stream.codebase_deploy_manifests.name
+    }
+  }
+
+  source {
+    type      = "CODEPIPELINE"
+    buildspec = templatefile("${path.module}/buildspec-manifests.yml", { application = var.application, environments = [for env in each.value.environments : upper(env.name)], services = local.service_export_names })
```
Comment (WillGibson) [inline codebase-pipelines/codebuild.tf:151]: Again wondering if a bit of multi-lining would make this easier to follow.

File: codebase-pipelines/artifactstore.tf [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  bucket = "${var.application}-${var.codebase}-codebase-pipeline-artifact-store"
+
+  tags = local.tags
+}
+
+data "aws_iam_policy_document" "artifact_store_bucket_policy" {
+  statement {
+    principals {
+      type        = "*"
+      identifiers = ["*"]
```
Comment (WillGibson) [inline codebase-pipelines/artifactstore.tf:16]: Could these be 
```suggestion
      type        = "Service"
      identifiers = ["codepipeline.amazonaws.com"]
```

File: codebase-pipelines/artifactstore.tf [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    actions = [
+      "s3:*",
```
[the remark this answers is not in the document]
Comment (JohnStainsby) [reply codebase-pipelines/artifactstore.tf:20]: This is a deny all policy to ensure SSL only access

File: codebase-pipelines/artifactstore.tf [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  policy = data.aws_iam_policy_document.artifact_store_bucket_policy.json
+}
+
+resource "aws_kms_key" "artifact_store_kms_key" {
+  # checkov:skip=CKV_AWS_7:We are not currently rotating the keys
+  description = "KMS Key for S3 encryption"
+  tags        = local.tags
+
+  policy = jsonencode({
+    Id = "key-default-1"
```
[the remark this answers is not in the document]
Comment (JohnStainsby) [reply codebase-pipelines/artifactstore.tf:52]: It's not required so we can probably get rid of it, we're using it everywhere for KMS key policies

File: codebase-pipelines/artifactstore.tf (hunk 3 above)
Comment (JohnStainsby) [reply codebase-pipelines/artifactstore.tf:16]: This is a deny all policy to ensure SSL only access

**Comment by JohnStainsby** [comment]
> A first pass. Brain is a bit too Friday afternoon for this many lines!
> 
> Have you been pairing on this? I'm wondering how deep I need to get into it to review it.

Yes, @james-francis-MT and I have been pairing on most of this. We've split some of it off into different PRs, but yeah it's big

File: codebase-pipelines/artifactstore.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (WillGibson) [reply codebase-pipelines/artifactstore.tf:36]: Ah, cool, I hadn't clocked the `effect = "Deny"`.

File: codebase-pipelines/tests/unit.tftest.hcl [hunk 6]
```diff
@@ -28,11 +28,58 @@ override_data {
...
+    json = "{\"Sid\": \"AssumeEventBridge\"}"
+  }
+}
+
+override_data {
+  target = data.aws_iam_policy_document.event_bridge_pipeline_trigger
+  values = {
+    json = "{\"Sid\": \"EventBridgePipelineTrigger\"}"
+  }
+}
```
Comment (WillGibson) [inline codebase-pipelines/tests/unit.tftest.hcl:64]: I know we cannot assert on the contents on a plan, but we should be able to assert on some basic things, like have we at least used the correct policy document...

https://github.com/uktrade/terraform-platform-modules/commit/1cd29ab7f4955166656306d4a3086f707ebf314d#diff-8494f4b8e2b47c589035d642f9c28ffa867e9c9a25dc0055f6b9fc4928796860

Comment (JohnStainsby) [reply codebase-pipelines/tests/unit.tftest.hcl:64]: I realised we can test more stuff. It's only policy documents with resources that need to be evaluated at apply that we can't tests.

So I've added a bunch of policy document tests and links between policies and roles where possible.

**Review by ejayesh — APPROVED** [review]
terraform code lgtm, looking forward to this going in.

**Pull request merged**